Sources/NetworkProtection/PacketTunnelProvider.swift

```diff
@@ -1670,6 +1690,97 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
             }
         }
     }
+
+    // MARK: - Snooze
+
+    private func startSnooze(_ duration: TimeInterval, completionHandler: ((Data?) -> Void)? = nil) {
+        Task {
+            await startSnooze(duration: duration)
+            completionHandler?(nil)
+        }
+    }
+
+    private func cancelSnooze(completionHandler: ((Data?) -> Void)? = nil) {
+        Task {
+            await cancelSnooze()
+            completionHandler?(nil)
+        }
+    }
+
+    private var snoozeTimer: DispatchSourceTimer?
+    private var snoozeRequestProcessing: Bool = false
+    private var snoozeJustEnded: Bool = false
+
+    @MainActor
+    private func startSnooze(duration: TimeInterval) async {
+        if snoozeRequestProcessing {
+            os_log("Rejecting start snooze request due to existing request processing", log: .networkProtection)
+            return
+        }
+
+        snoozeRequestProcessing = true
+        os_log("Starting snooze mode with duration: %{public}d", log: .networkProtection, duration)
+
+        await stopMonitors()
+
+        self.adapter.snooze { [weak self] error in
+            guard let self else {
+                assertionFailure("Failed to get strong self")
+                return
+            }
+
+            if error == nil {
+                self.connectionStatus = .snoozing
+                self.snoozeTimingStore.activeTiming = .init(startDate: Date(), duration: duration)
+                self.notificationsPresenter.showSnoozingNotification(duration: duration)
+
+                let timer = DispatchSource.makeTimerSource(queue: self.timerQueue)
+                timer.schedule(deadline: .now() + duration, repeating: .never)
+                timer.setEventHandler {
+                    Task {
+                        os_log("Snooze mode timer expired, canceling snooze now...", log: .networkProtection)
+                        await self.cancelSnooze()
+                    }
+
+                    timer.cancel()
+                }
+
+                self.snoozeTimer = timer
+                self.snoozeTimer?.resume()
+            } else {
+                self.snoozeTimingStore.reset()
```

Comment:
Shouldn't we do something here, like firing an error pixel?

Reply:
The one error that can occur is if the adapter is already stopped, which is an error we're not too concerned about; it could happen if a snooze call comes in right as the VPN is shutting down, which isn't too interesting. I'll likely come back and add a pixel here to be safe though, after triaging it first.